services/requester/keystore/key_store.go

```diff
@@ -14,30 +17,57 @@ const accountKeyBlockExpiration = flow.DefaultTransactionExpiry
 
 type KeyLock interface {
 	NotifyTransaction(txID flowsdk.Identifier)
-	NotifyBlock(blockHeight uint64)
+	NotifyBlock(blockID flowsdk.Identifier)
```

Comment:
Can we add comments to these two methods for scenarios they will be used?

And since this PR is to address the swam of invalid transactions (nonce too lower), I understand it takes time for NotifyBlockByHeight to release them, but why can't GW release them through NotifyTransaction? I think those invalid transactions are all included in the blocks, therefore will be seen and indexed by GW locally, no?

Reply:
Good idea :+1: I added some descriptive comments in https://github.com/onflow/flow-evm-gateway/pull/842/commits/7f0145fc866233855239c097b9156a7ff608410c .

> I think those invalid transactions are all included in the blocks, therefore will be seen and indexed by GW locally, no?

You are right, they are generally included in Flow blocks. But we use the Event Streaming API in our event ingestion, and we only subscribe to:
- `EVM.TransactionExecuted`
- `EVM.BlockExecuted`

events. And invalid EVM transactions do not emit `EVM.TransactionExecuted` events. That's why we resort to the gRPC API, `GetTransactionResultsByBlockID`.

Reply:
> Or is it possible for An to add a new event such as EVM.TransactionRejected events for invalid transactions? Then we don't even need to fetch tx IDs for each block.

Cannot emit event there as flow transaction panics.

Reply:
We had discussed the possibility of using an EVM event dedicated for this (such as `EVM.TransactionRejected`), but it has certain drawbacks:
- Events are stored on-chain
- Trading bots are spamming 5-10 such transactions per Flow block, so about 5-10 such EVM events will be emitted per block. And that only reflects the activity as of today.
- Events in general are supposed to give insights about on-chain state changes. In this case, we're introducing an event for a technicality, which doesn't provide any actual insight to devs/users.

In general, I think we're better off using an off-chain approach for the issue at hand.

Reply:
@peterargue  But shouldn't delayed keep the sequence number but others increase? 

Only issue maybe can be some AN falling back somehow I guess. But as seq never decrease shouldn't be that much of a deal breaker.